lib/vector-core/src/transform/mod.rs

```diff
@@ -224,23 +244,37 @@ async fn send_inner(buf: &mut Vec<Event>, output: &mut Fanout) {
 }
 
 pub struct TransformOutputsBuf {
-    primary_buffer: Vec<Event>,
+    primary_buffer: Option<Vec<Event>>,
     named_buffers: HashMap<String, Vec<Event>>,
 }
 
 impl TransformOutputsBuf {
-    pub fn new_with_capacity(named_outputs_in: Vec<String>, capacity: usize) -> Self {
+    pub fn new_with_capacity(outputs_in: Vec<Output>, capacity: usize) -> Self {
+        let mut primary_buffer = None;
+        let mut named_buffers = HashMap::new();
+
+        for output in outputs_in {
+            match output.port {
+                None => {
+                    primary_buffer = Some(Vec::with_capacity(capacity));
+                }
+                Some(name) => {
+                    named_buffers.insert(name.clone(), Vec::new());
```

Comment:
not sure if it'd improve the performance but do you think at some point we could have a lifetime for the `named_buffers` key like
```rust
pub struct TransformOutputsBuf<'key> {
    ...
    named_buffers: HashMap<&'key String, Vec<Event>>,
}
```
So that we avoid cloning string for each events we receive?

Reply:
Yeah that's an interesting idea. It's currently going to clone once per batch of events, so not quite so bad as once per event. I had planned to continue working on the efficiency of the internals of both this and `SourceSender`, and having some kind of lifetime associated with the parent `TransformOutputs` would be an interesting idea.